### PR TITLE
Feat/add npm ci

### PR DESCRIPTION
**Pull request by julien-devatom**
## Motivation
Morpho blue does not have any dependency (forge or npm). We can use the contracts in both `hardhat` and `foundry` projects. 
If we want to ease Blue integrations to devs not working on Foundry, we can publish the package to npm.

However, publication to npm is a critical action that needs to have security about when and who can trigger the CI.

## Solution
- The workflow can be triggered by hand.
- 've not defined an auto versioning not to have to allow the action to push on main the bumped version (security reason).
- However, it is up to the person triggering the workflow to define a correct version. 
- There is a check to verify the semver format.
- I have created a GitHub environment (named npm) that is scoping the deployment to the main branch only and that requires administrators' reviewals before publishing on npm

### CODE REVIEW

**Review by MathisGD — DISMISSED** [review]

File: package.json
Comment (Rubilmax) [inline package.json:5]: Resources:
- https://docs.npmjs.com/cli/v10/configuring-npm/package-json#version
- https://github.com/npm/node-semver

It seems this version **might** work. Why aren't we using `1.0.0` instead?

File: package.json
Comment (Rubilmax) [inline package.json:10]: Can we clarify why `lib` is included in the package bundle before merging this pls?

File: package.json
Comment (Rubilmax) [reply package.json:5]: Is it the syntax of `yarn version`??

Comment (julien-devatom) [reply package.json:5]: Yes, this is the default syntax used by version. Here we are using hardcoded version, but I've reused the default behavior of --patch, --minor & --major

File: package.json
Comment (julien-devatom) [reply package.json:10]: A bit weird since its working well on this repo..

image

File: package.json
Comment (Rubilmax) [reply package.json:5]: If I do it myself, I can increment versions using `yarn version --patch/minor/major` and it will keep the syntax `x.x.x` instead of `vx.x.x`

File: .github/workflows/npm-release.yml
Comment (Rubilmax) [inline .github/workflows/npm-release.yml:33]: The `--new-version` parameter seems to be available with `yarn publish`: https://classic.yarnpkg.com/lang/en/docs/cli/publish/#toc-yarn-publish-new-version

File: .github/workflows/npm-release.yml
Comment (Rubilmax) [inline .github/workflows/npm-release.yml:6]: It seems you expect as input a specific version and not a version type (patch,minor,major). Is it expected? Why not use incremental versioning, considering versions cannot be republished on npm?

File: .github/workflows/npm-release.yml
Comment (Rubilmax) [reply .github/workflows/npm-release.yml:33]: Just tested it: it doesn't commit it to git
But it's actually a benefit because it doesn't break the commit msg syntax convention that we use (whereas `yarn version` commits `vx.x.x` as commit)
So it means using `yarn publish --new-version` allows to publish the new version and optionally commit it with a defined commit msg

Also, committing the version **after** having published it seems better because in case the `publish` step fails, no git tag/commit is pushed

**Review by Rubilmax — CHANGES_REQUESTED** [review]

File: package.json
Comment (Rubilmax) [reply package.json:10]: You don't have to publish a package to npm to know this ahah: `yarn pack`

Comment (Rubilmax) [reply package.json:10]: Alright I confirm so this discussion can be resolved

**Comment by Rubilmax** [comment]
Just in case we ever publish ts: https://github.com/morpho-org/morpho-blue/pull/561

File: .github/workflows/npm-release.yml
Comment (Rubilmax) [reply .github/workflows/npm-release.yml:33]: I am surprised they recommend to push the commit right after versioning: https://classic.yarnpkg.com/lang/en/docs/cli/version/#toc-version-lifecycle

But I guess it only depends on what we want. So no strict requirement on this from my end in the end

I also discover: `yarn config set version-git-message` which I think is important though, for our commit msg convention

**Review by Rubilmax — CHANGES_REQUESTED** [review]

File: package.json
Comment (julien-devatom) [reply package.json:10]: Wanted to reproduce the exact same flow (idk if publish has some extra things that pack has not), btw Ive just deleted the package 😆

File: .github/workflows/npm-release.yml
Comment (julien-devatom) [reply .github/workflows/npm-release.yml:6]: This is written in the PR description: 've not defined an auto versioning not to have to allow the action to push on main the bumped version (security reason).

Autoincrement is based on the version provided inside of the package.json, so the CI need to push the bumped version on main, which is not a good thing for a critical repo like this one IMO.

File: .github/workflows/npm-release.yml
Comment (Rubilmax) [reply .github/workflows/npm-release.yml:33]: Isn't there a way to tell `yarn` to not version anything and just use the package's version defined in `package.json`?

So the flow would be:
1. commit a new version to `package.json`
2. merge
3. trigger action without any input
4. if action is triggered twice with the same version, fail

Comment (julien-devatom) [reply .github/workflows/npm-release.yml]: That's a good tradeoff. Ive also reduced the version to 0.1.0, and we'll release the 1.0.0 once deployed (it should never change after)

**Review by Rubilmax — APPROVED** [review]

File: .github/workflows/npm-release.yml
Comment (Rubilmax) [reply .github/workflows/npm-release.yml]: This discussion can be resolved but I keep it open for others to know

**Review by MerlinEgalite — APPROVED** [review]

**Review by Jean-Grimal — APPROVED** [review]

**Pull request merged**
